packages/ember-data/lib/serializers/rest-serializer.js

```diff
@@ -683,12 +714,72 @@ var RESTSerializer = JSONSerializer.extend({
   serializePolymorphicType: function(snapshot, json, relationship) {
     var key = relationship.key;
     var belongsTo = snapshot.belongsTo(key);
+    var typeKey = this.keyForPolymorphicType(key, relationship.type, 'serialize');
+
+    // old way of getting the key for the polymorphic type
     key = this.keyForAttribute ? this.keyForAttribute(key, "serialize") : key;
+    key = `${key}Type`;
+
+    // check if the old way differs with the new, correct one, and if so, log a
+    // deprecation warning and restore the old behavior
+    if (key !== typeKey) {
+      Ember.deprecate("The typeKey to serialize a polymorphic type uses keyForAttribute which has been deprecated in favor of keyForPolymorphicType.", {
+        id: 'ds.rest-serializer.deprecated-key-for-polymorphic-type',
+        until: '3.0.0'
+      });
+
+      typeKey = key;
+    }
```

Comment:
The current way to get the `myKeyType` string for the `myKey` relationship is using `keyForAttribute` which is not correct IMHO. This is changed in the course of this PR by adding a `keyForPolymorphicType` hook.

This could break current behavior, that's why a deprecation message is added. I will add a test for the deprecation if the deprecation is found to be necessary.